File: install/thor_teleop/lib/thor_teleop/teleop_pose_control.py
#!/usr/bin/env python3
# ========================================================================
# File: thor_teleop/scripts/teleop_pose_control.py
# ========================================================================
# This node provides keyboard teleoperation for the end-effector's
# Cartesian pose (position and orientation).
#
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
import threading
import sys
import termios
import tty
import math
import time

from std_msgs.msg import Bool
from geometry_msgs.msg import PoseStamped, Pose, Quaternion
from tf2_ros import Buffer, TransformListener
from tf_transformations import quaternion_from_euler, quaternion_multiply

from moveit_msgs.action import MoveGroup
from moveit_msgs.msg import MotionPlanRequest, Constraints, PositionConstraint, OrientationConstraint
from shape_msgs.msg import SolidPrimitive

# --- Key Mapping and Configuration ---
KEY_MAP = {
    'w': ('x', 0.02), 's': ('x', -0.02),  # Forward/Backward
    'a': ('y', 0.02), 'd': ('y', -0.02),  # Left/Right
    'r': ('z', 0.02), 'f': ('z', -0.02),  # Up/Down
    'q': ('roll', 0.1), 'e': ('roll', -0.1),
    't': ('pitch', 0.1), 'g': ('pitch', -0.1),
    'y': ('yaw', 0.1), 'h': ('yaw', -0.1),
}

class TeleopPoseControl(Node):
    def __init__(self):
        super().__init__('teleop_pose_control')

        # --- Parameters ---
        self.move_group_name = "arm_group"
        self.end_effector_link = "electromagnet_plate"
        self.base_frame = "world"

        # --- State ---
        self.current_pose = None
        self.electromagnet_state = False
        self.motion_in_progress = False
        self.state_lock = threading.Lock()

        # --- ROS Communications ---
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.move_group_client = ActionClient(self, MoveGroup, '/move_action')
        self.electromagnet_pub = self.create_publisher(Bool, '/thor_arm/electromagnet/control', 10)
        
        self.get_logger().info("🤖 Thor Pose (Cartesian) Teleop Controller Started.")
        self.print_instructions()

        # Start keyboard listener thread
        self.keyboard_thread = threading.Thread(target=self.keyboard_loop)
        self.keyboard_thread.daemon = True
        self.keyboard_thread.start()

    def print_instructions(self):
        """Prints the control instructions to the console."""
        print("\n" + "="*50)
        print("        Thor Pose (Cartesian) Teleop Control")
        print("="*50)
        print("  Position (XYZ):")
        print("    Fwd/Bwd [w]/[s]   Left/Right [a]/[d]   Up/Down [r]/[f]")
        print("\n  Orientation (Roll, Pitch, Yaw):")
        print("    Roll: [q]/[e]   Pitch: [t]/[g]   Yaw: [y]/[h]")
        print("\n  End Effector:")
        print("    [spacebar]: Toggle Electromagnet")
        print("\n  [Ctrl+C] to quit")
        print("="*50 + "\n")

    def keyboard_loop(self):
        """Continuously listens for keyboard input."""
        self.get_logger().info("Waiting for MoveIt server...")
        self.move_group_client.wait_for_server()
        self.get_logger().info("✅ MoveIt server is ready.")

        # Get initial robot pose
        self.get_current_pose()

        while rclpy.ok():
            key = self.get_key()
            if key == '\x03': # Ctrl+C
                break
            
            with self.state_lock:
                if self.motion_in_progress:
                    self.get_logger().warn("⚠️ Motion in progress, please wait.")
                    continue
                if self.current_pose is None:
                    self.get_logger().warn("⚠️ Current pose not yet available.")
                    continue
                self.motion_in_progress = True

            if key in KEY_MAP:
                axis, increment = KEY_MAP[key]
                self.update_pose(axis, increment)
                self.move_to_pose(self.current_pose)
            elif key == ' ':
                self.toggle_electromagnet()
                self.motion_in_progress = False # No motion, release lock
            else:
                self.motion_in_progress = False # No valid key, release lock

    def get_key(self):
        """Gets a single key press from the terminal."""
        fd = sys.stdin.fileno()
        old_settings = termios.tcgetattr(fd)
        try:
            tty.setraw(sys.stdin.fileno())
            key = sys.stdin.read(1)
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
        return key

    def get_current_pose(self):
        """Retrieves the current end-effector pose from TF2."""
        self.get_logger().info("Getting current end-effector pose...")
        # Clear the current pose to ensure we get a fresh one
        self.current_pose = None 
        while self.current_pose is None and rclpy.ok():
            try:
                transform_stamped = self.tf_buffer.lookup_transform(
                    self.base_frame, self.end_effector_link, rclpy.time.Time()
                )
                
                pose = Pose()
                pose.position.x = transform_stamped.transform.translation.x
                pose.position.y = transform_stamped.transform.translation.y
                pose.position.z = transform_stamped.transform.translation.z
                pose.orientation = transform_stamped.transform.rotation
                
                self.current_pose = PoseStamped(header=transform_stamped.header, pose=pose)
                
                self.get_logger().info(f"✅ Pose acquired: "
                                       f"P({self.current_pose.pose.position.x:.2f}, "
                                       f"{self.current_pose.pose.position.y:.2f}, "
                                       f"{self.current_pose.pose.position.z:.2f})")
            except Exception as e:
                self.get_logger().warn(f"Could not get transform, retrying: {e}")
                time.sleep(1.0)

    def update_pose(self, axis, increment):
        """Updates the target pose based on the key press."""
        p = self.current_pose.pose.position
        o = self.current_pose.pose.orientation
        
        if axis == 'x': p.x += increment
        elif axis == 'y': p.y += increment
        elif axis == 'z': p.z += increment
        elif axis in ['roll', 'pitch', 'yaw']:
            delta_q = {
                'roll': quaternion_from_euler(increment, 0, 0),
                'pitch': quaternion_from_euler(0, increment, 0),
                'yaw': quaternion_from_euler(0, 0, increment),
            }[axis]
            
            current_q = [o.x, o.y, o.z, o.w]
            new_q = quaternion_multiply(current_q, delta_q)
            o.x, o.y, o.z, o.w = new_q[0], new_q[1], new_q[2], new_q[3]
        
        self.get_logger().info(f"➡️ New Target: P({p.x:.2f}, {p.y:.2f}, {p.z:.2f})")

    def toggle_electromagnet(self):
        """Toggles the state of the electromagnet."""
        self.electromagnet_state = not self.electromagnet_state
        msg = Bool(data=self.electromagnet_state)
        self.electromagnet_pub.publish(msg)
        self.get_logger().info(f"🧲 Electromagnet toggled {'ON' if self.electromagnet_state else 'OFF'}")

    def move_to_pose(self, pose_stamped):
        """Sends a goal to move to a target pose."""
        goal_msg = MoveGroup.Goal()
        request = MotionPlanRequest()
        request.group_name = self.move_group_name
        request.num_planning_attempts = 5
        request.allowed_planning_time = 5.0
        
        constraints = Constraints()
        
        pos_constraint = PositionConstraint()
        pos_constraint.header.frame_id = self.base_frame
        pos_constraint.link_name = self.end_effector_link
        pos_constraint.constraint_region.primitives.append(SolidPrimitive(type=SolidPrimitive.SPHERE, dimensions=[0.01])) # 1cm tolerance
        pos_constraint.constraint_region.primitive_poses.append(pose_stamped.pose)
        pos_constraint.weight = 1.0
        constraints.position_constraints.append(pos_constraint)
        
        orient_constraint = OrientationConstraint()
        orient_constraint.header.frame_id = self.base_frame
        orient_constraint.link_name = self.end_effector_link
        orient_constraint.orientation = pose_stamped.pose.orientation
        orient_constraint.absolute_x_axis_tolerance = 0.2
        orient_constraint.absolute_y_axis_tolerance = 0.2
        orient_constraint.absolute_z_axis_tolerance = 0.2
        orient_constraint.weight = 1.0
        constraints.orientation_constraints.append(orient_constraint)

        request.goal_constraints.append(constraints)
        goal_msg.request = request
        
        future = self.move_group_client.send_goal_async(goal_msg)
        future.add_done_callback(self.goal_done_callback)

    def goal_done_callback(self, future):
        """Callback for when a motion is complete."""
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().error('❌ Goal rejected by MoveIt server')
            with self.state_lock:
                self.motion_in_progress = False
        else:
            result_future = goal_handle.get_result_async()
            result_future.add_done_callback(self.result_callback)

    def result_callback(self, future):
        """Callback for the final result of the motion."""
        result = future.result().result
        if result.error_code.val == 1: # SUCCESS
            self.get_logger().info("✅ Motion succeeded.")
        else:
            self.get_logger().error(f"❌ Motion failed with error code: {result.error_code.val}")
        
        # *** FIX: Always re-acquire the pose from TF after a move attempt ***
        # This ensures the next jog starts from the robot's true position.
        self.get_logger().info("Re-acquiring current pose from TF to re-sync.")
        self.get_current_pose()

        with self.state_lock:
            self.motion_in_progress = False

def main(args=None):
    rclpy.init(args=args)
    node = TeleopPoseControl()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/thor_teleop/src/teleop_pose_control.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <thread>
#include <mutex>
#include <atomic>
#include <termios.h>
#include <unistd.h>

#include "std_msgs/msg/bool.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Matrix3x3.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"

#include "moveit_msgs/action/move_group.hpp"
#include "moveit_msgs/msg/motion_plan_request.hpp"
#include "moveit_msgs/msg/constraints.hpp"
#include "moveit_msgs/msg/position_constraint.hpp"
#include "moveit_msgs/msg/orientation_constraint.hpp"
#include "shape_msgs/msg/solid_primitive.hpp"

// Key mapping for pose control
const std::map<char, std::pair<std::string, double>> KEY_MAP = {
    {'w', {"x", 0.02}}, {'s', {"x", -0.02}},
    {'a', {"y", 0.02}}, {'d', {"y", -0.02}},
    {'r', {"z", 0.02}}, {'f', {"z", -0.02}},
    {'q', {"roll", 0.1}}, {'e', {"roll", -0.1}},
    {'t', {"pitch", 0.1}}, {'g', {"pitch", -0.1}},
    {'y', {"yaw", 0.1}}, {'h', {"yaw", -0.1}}
};

class TeleopPoseControl : public rclcpp::Node
{
public:
    TeleopPoseControl() : Node("teleop_pose_control")
    {
        // ROS Communications
        move_group_client_ = rclcpp_action::create_client<moveit_msgs::action::MoveGroup>(this, "/move_action");
        electromagnet_pub_ = this->create_publisher<std_msgs::msg::Bool>("/thor_arm/electromagnet/control", 10);
        
        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        RCLCPP_INFO(this->get_logger(), "🤖 Thor Pose (Cartesian) Teleop Controller Started.");
        print_instructions();

        // Start keyboard listener in a separate thread
        keyboard_thread_ = std::thread(&TeleopPoseControl::keyboard_loop, this);
    }

    ~TeleopPoseControl()
    {
        if (keyboard_thread_.joinable())
        {
            keyboard_thread_.join();
        }
    }

private:
    // ROS Components
    rclcpp_action::Client<moveit_msgs::action::MoveGroup>::SharedPtr move_group_client_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr electromagnet_pub_;
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

    // State
    std::shared_ptr<geometry_msgs::msg::PoseStamped> current_pose_;
    std::atomic<bool> motion_in_progress_{false};
    std::atomic<bool> electromagnet_state_{false};
    std::mutex pose_mutex_;
    std::thread keyboard_thread_;

    // Parameters
    const std::string move_group_name_ = "arm_group";
    const std::string end_effector_link_ = "electromagnet_plate";
    const std::string base_frame_ = "world";

    void print_instructions()
    {
        RCLCPP_INFO(this->get_logger(), "\n"
            "==================================================\n"
            "        Thor Pose (Cartesian) Teleop Control\n"
            "==================================================\n"
            "  Position (XYZ):\n"
            "    Fwd/Bwd [w]/[s]   Left/Right [a]/[d]   Up/Down [r]/[f]\n"
            "\n"
            "  Orientation (Roll, Pitch, Yaw):\n"
            "    Roll: [q]/[e]   Pitch: [t]/[g]   Yaw: [y]/[h]\n"
            "\n"
            "  End Effector:\n"
            "    [spacebar]: Toggle Electromagnet\n"
            "\n"
            "  [Ctrl+C] to quit\n"
            "==================================================");
    }

    char get_key()
    {
        struct termios old_tio, new_tio;
        tcgetattr(STDIN_FILENO, &old_tio);
        new_tio = old_tio;
        new_tio.c_lflag &= (~ICANON & ~ECHO);
        tcsetattr(STDIN_FILENO, TCSANOW, &new_tio);

        char c = getchar();

        tcsetattr(STDIN_FILENO, TCSANOW, &old_tio);
        return c;
    }

    void keyboard_loop()
    {
        RCLCPP_INFO(this->get_logger(), "Waiting for MoveIt server...");
        move_group_client_->wait_for_action_server();
        RCLCPP_INFO(this->get_logger(), "✅ MoveIt server is ready.");

        get_current_pose();

        while (rclcpp::ok())
        {
            char key = get_key();
            if (key == '\x03') // Ctrl+C
            {
                rclcpp::shutdown();
                break;
            }

            if (motion_in_progress_.load())
            {
                RCLCPP_WARN(this->get_logger(), "⚠️ Motion in progress, please wait.");
                continue;
            }

            {
                std::lock_guard<std::mutex> lock(pose_mutex_);
                if (!current_pose_)
                {
                    RCLCPP_WARN(this->get_logger(), "⚠️ Current pose not yet available.");
                    continue;
                }
            }
            
            motion_in_progress_.store(true);

            if (KEY_MAP.count(key))
            {
                const auto& [axis, increment] = KEY_MAP.at(key);
                update_pose(axis, increment);
                move_to_pose();
            }
            else if (key == ' ')
            {
                toggle_electromagnet();
                motion_in_progress_.store(false);
            }
            else
            {
                motion_in_progress_.store(false);
            }
        }
    }

    void get_current_pose()
    {
        RCLCPP_INFO(this->get_logger(), "Getting current end-effector pose...");
        {
            std::lock_guard<std::mutex> lock(pose_mutex_);
            current_pose_ = nullptr;
        }

        while (!current_pose_ && rclcpp::ok())
        {
            try
            {
                geometry_msgs::msg::TransformStamped t = tf_buffer_->lookupTransform(
                    base_frame_, end_effector_link_, tf2::TimePointZero);
                
                std::lock_guard<std::mutex> lock(pose_mutex_);
                current_pose_ = std::make_shared<geometry_msgs::msg::PoseStamped>();
                current_pose_->header = t.header;
                current_pose_->pose.position.x = t.transform.translation.x;
                current_pose_->pose.position.y = t.transform.translation.y;
                current_pose_->pose.position.z = t.transform.translation.z;
                current_pose_->pose.orientation = t.transform.rotation;

                RCLCPP_INFO(this->get_logger(), "✅ Pose acquired: P(%.2f, %.2f, %.2f)",
                    current_pose_->pose.position.x, current_pose_->pose.position.y, current_pose_->pose.position.z);
            }
            catch (const tf2::TransformException & ex)
            {
                RCLCPP_WARN(this->get_logger(), "Could not get transform, retrying: %s", ex.what());
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
        }
    }

    void update_pose(const std::string& axis, double increment)
    {
        std::lock_guard<std::mutex> lock(pose_mutex_);
        auto& p = current_pose_->pose.position;
        auto& o = current_pose_->pose.orientation;

        if (axis == "x") p.x += increment;
        else if (axis == "y") p.y += increment;
        else if (axis == "z") p.z += increment;
        else if (axis == "roll" || axis == "pitch" || axis == "yaw")
        {
            tf2::Quaternion current_q;
            tf2::fromMsg(o, current_q);

            tf2::Quaternion delta_q;
            if (axis == "roll") delta_q.setRPY(increment, 0, 0);
            if (axis == "pitch") delta_q.setRPY(0, increment, 0);
            if (axis == "yaw") delta_q.setRPY(0, 0, increment);

            current_q = current_q * delta_q;
            current_q.normalize();
            o = tf2::toMsg(current_q);
        }
        RCLCPP_INFO(this->get_logger(), "➡️ New Target: P(%.2f, %.2f, %.2f)", p.x, p.y, p.z);
    }

    void toggle_electromagnet()
    {
        electromagnet_state_ = !electromagnet_state_.load();
        auto msg = std_msgs::msg::Bool();
        msg.data = electromagnet_state_.load();
        electromagnet_pub_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "🧲 Electromagnet toggled %s", electromagnet_state_.load() ? "ON" : "OFF");
    }

    void move_to_pose()
    {
        auto goal_msg = moveit_msgs::action::MoveGroup::Goal();
        auto& request = goal_msg.request;
        request.group_name = move_group_name_;
        request.num_planning_attempts = 5;
        request.allowed_planning_time = 5.0;

        moveit_msgs::msg::Constraints constraints;
        
        // Position Constraint
        shape_msgs::msg::SolidPrimitive sphere;
        sphere.type = shape_msgs::msg::SolidPrimitive::SPHERE;
        sphere.dimensions.push_back(0.01); // 1cm radius

        moveit_msgs::msg::PositionConstraint pos_constraint;
        pos_constraint.header.frame_id = base_frame_;
        pos_constraint.link_name = end_effector_link_;
        pos_constraint.constraint_region.primitives.push_back(sphere);
        {
            std::lock_guard<std::mutex> lock(pose_mutex_);
            pos_constraint.constraint_region.primitive_poses.push_back(current_pose_->pose);
        }
        pos_constraint.weight = 1.0;
        constraints.position_constraints.push_back(pos_constraint);

        // Orientation Constraint
        moveit_msgs::msg::OrientationConstraint orient_constraint;
        orient_constraint.header.frame_id = base_frame_;
        orient_constraint.link_name = end_effector_link_;
        {
            std::lock_guard<std::mutex> lock(pose_mutex_);
            orient_constraint.orientation = current_pose_->pose.orientation;
        }
        orient_constraint.absolute_x_axis_tolerance = 0.2;
        orient_constraint.absolute_y_axis_tolerance = 0.2;
        orient_constraint.absolute_z_axis_tolerance = 0.2;
        orient_constraint.weight = 1.0;
        constraints.orientation_constraints.push_back(orient_constraint);

        request.goal_constraints.push_back(constraints);
        
        using namespace std::placeholders;
        auto send_goal_options = rclcpp_action::Client<moveit_msgs::action::MoveGroup>::SendGoalOptions();
        send_goal_options.goal_response_callback = std::bind(&TeleopPoseControl::goal_response_callback, this, _1);
        send_goal_options.result_callback = std::bind(&TeleopPoseControl::result_callback, this, _1);
        
        move_group_client_->async_send_goal(goal_msg, send_goal_options);
    }

    void goal_response_callback(const rclcpp_action::ClientGoalHandle<moveit_msgs::action::MoveGroup>::SharedPtr& goal_handle)
    {
        if (!goal_handle) {
            RCLCPP_ERROR(this->get_logger(), "❌ Goal was rejected by server");
            motion_in_progress_.store(false);
        } else {
            RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for result...");
        }
    }

    void result_callback(const rclcpp_action::ClientGoalHandle<moveit_msgs::action::MoveGroup>::WrappedResult& result)
    {
        if (result.code == rclcpp_action::ResultCode::SUCCEEDED) {
            RCLCPP_INFO(this->get_logger(), "✅ Motion succeeded.");
        } else {
            RCLCPP_ERROR(this->get_logger(), "❌ Motion failed with error code: %d", result.result->error_code.val);
        }

        RCLCPP_INFO(this->get_logger(), "Re-acquiring current pose from TF to re-sync.");
        get_current_pose();
        motion_in_progress_.store(false);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TeleopPoseControl>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
